### DISCUSSION TITLE
As a content creator, I would like to create pages using an admin portal.

### DISCUSSION THREAD

**Discussion by Bowens20832** [elegantframework/elegant-cli#49] [RFC] [answered]
## Background / Motivation

Currently if you want to create new pages or content within Elegant, you have to manually create new Markdown pages.

I, myself am planning on adding a basic CMS editor when some of the more basic v2.* features have been completed.

Having an editor editor panel will allow creators to create new pages much more rapidly, as well as opening the door up for non technical users to being using Elegant for their website needs.

[@rabisnaqvi](https://github.com/rabisnaqvi) has also suggested including an admin panel/portal in #44.

## Idea

Vercel already has a number of documented basic editor panels that can be integrated into Next.js, so I'd like to not recreate anything we don't have to.

[Tina](https://github.com/vercel/next.js/tree/canary/examples/cms-tina
)
## Implementation 1:

Use an open source CMS panel, or create our own if we can't an existing solution.

Just use Tina if we can self host, not be attached to any of their services, and extend their existing code.

## Other things

Can we build a simple CMS panel that can then be hosted easily and securly on Vercel or Netlify?

Do we need to include an auth provider like Firebase or Next-Auth?

**Comment by Bowens20832** [comment]
Does anyone have any experience with Tina CMS? [https://github.com/vercel/next.js/tree/canary/examples/cms-tina](https://github.com/vercel/next.js/tree/canary/examples/cms-tina)

**Comment by rabisnaqvi** [comment]
Tina seems promising. I think before we dive into the whole CMS ordeal, we need to first define how we want to tackle the database aspects of the framework.

- Do we want to require a database to be present and integrated?
- Do we save the pages, and other data types in DB or as files in the namespaces?
- Do we want to have a predefined login and authentication system?
- Do we separate the core and the portal applications? for example, the core elegant framework could be totally dev dependent and would need configuration, etc to work, but when combined with the portal, it could work as a complete end-to-end framework without the need for initial or boilerplate config.
- if we use a database, which system do we use? or do we want to support multiple types of databases? which tools and ORMs do we use?

These are some questions that I believe we should brainstorm on first. @Bowens20832

**Reply by Bowens20832** [reply]
I have been brainstorming the database piece myself, and talking with developers and potential users.

I believe everything we could want is covered in the Vercel guide: https://vercel.com/guides/nextjs-multi-tenant-application

- Yes, the market seems to really want a WordPress type CMS, but with a serverless database instead of a hosted database as WordPress does now. PlanetScale, SupaBase, and Vercel Postgres all look promising.

- I'm thinking we can leverage the power of static pages as we have now, and build a hybrid solution around Next.Js.

- Yes, Vercel and others have created serverless Auth packages that integrate seamlessly with Next.js and can authenticate with just about any service. Next Auth looks to be really promising: https://next-auth.js.org/

- I'm not sure about seperating the backend from the core code yet. I am planning on building out the Elegant-CLI tool to ask users questions during project creation, such as if they want the admin panel, and the cli tool will auto include the packages they want, as well as fill out their .env file for them.

- I'd like to stay flexible and write the code to support any type of DB the user may want to use. I'm thinking that developers would love a Laravel type frontend where they can pick their data storage method, or help us code up the database method they need.

What do you think? @rabisnaqvi

**Reply by rabisnaqvi** [reply]
- I personally have had a great experience with Supabase, appreciating its smooth handling and control, so I'm definitely in favor of that!
- Using static files as pages and routes may be straightforward but could lead to version control issues. I'm unsure if you plan to stick to static content or consider making it dynamic for/with potential extensions like WordPress.
- Integration of next-auth with Supabase for authentication could be a good idea.
- Yup, A CLI tool generating boilerplate code could help kickstart the framework.
- I believe instead of providing support for multiple database architectures initially, integrating Supabase or allowing users to connect their own instances of Supabase could be an easy starting point.

I have a question regarding Elegant's role. I'm uncertain whether it will function primarily as a development framework or an application development toolkit. In my understanding, WordPress leans more towards being an application development toolkit due to its emphasis on a no-code or low-code approach. On the other hand, a framework typically targets developers, offering greater flexibility but potentially introducing complexity and barriers to entry. I'm curious if you have a preference towards one end of the spectrum or if you're aiming to strike a balance between the two approaches.

🍻 @Bowens20832

**Reply by Bowens20832** [reply]
Yes, I believe we could strike a balance between both.

Both content creators and developers agree that WordPress and others are dated, and wish they had a better solution. 

At the same time, Next.js does almost everything anyone could want.

 I did some user surveys for feedback and most people didn't understand Next.js or the other serverless concepts. However, they did say they would be interested in using a modern serverless WordPress concept built with Next.js. 
 
 I am thinking that the long term outlook for Elegant is to continue the free open-source self hosted version, and eventually offer a cheap and easy to use cloud hosted version with no setup on the users end.

**Reply by Bowens20832** [reply]
@rabisnaqvi 
![MicrosoftTeams-image (6)](https://github.com/elegantframework/elegant-cli/assets/10189130/db6e3702-18ec-428e-ad76-2bb4f0401bf1)

**Reply by rabisnaqvi** [reply]
Here is my understanding, and a few questions based on your points:

CLI:
- Handles initialization, configuration, and module management.
- Generates boilerplate code and sets up the project based on user input.
- Fetches the latest core package upon request (not included in the CLI itself).
- Downloads and initializes modules/extensions or themes as specified during initialization.

Core module:
- Forms the main functionality built on Next.js.
- Utilizes markdown formatted static files.
- Allows extension and customization of functionality.
- Supports custom post types or namespaces for different content types.
- Provides lifecycle hooks for modules and themes to integrate with the core framework.
- Comes with a default theme.

Database module:
- Manages data storage when required.
- If cloud databases with authentication features are used, they handle or integrate with the authentication module.
- Question: Are the database modules used only for storing users, or do we also consider using them as cloud storage buckets?
- Question: Are different user roles supported?

CMS module:
- Enables the creation of data type namespaces using a graphical user interface (GUI).
- Supports content creation in various data types using a WYSIWYG editor.
- Generates markdown files and organizes them within the specified namespace.
- Possibly includes features like module and theme installation, configuration, and uninstallation.
- Allows website user management (if authentication module is present).
- Question: Should the authentication system be required for production use, allowing changes without requiring a release, or is it intended as a development tool only like storybook?

Modules and extensions:
- Stored in a separate directory or managed via npm and a configuration file.
- Modules hook into the core framework's lifecycle to modify or add core behavior.
- Themes provide the user interface for displaying content.

These points reflect my understanding of the system. Please let me know if you disagree with any of these points. By the way, do you have any plans to upgrade this to Next.js 13?

@Bowens20832

**Reply by Bowens20832** [reply]
Yes to all of the above! I believe we are on the same page with everything 😄 

Also yes to Next.js 13... All of the packages need to be upgraded, and I was planning on attempting to conquer that for version 3 since there might be breaking changes. 

React, Next, Node, and others can be upgraded, but I was having issues with the Markdown loading when I upgraded to the latest version of React, so I abandoned it for now.

**Reply by Bowens20832** [reply]
@rabisnaqvi I just submitted #84 to upgrade Elegant to Node 18 and React 18... Should I consider ugrading to Next 13 now?

**Comment by Bowens20832** [comment]
@danoely, we are discussing possible data storage methods, as well as auth methods.

Do you have any comments toward Planet Scale or Serverless functions within Vercel?

**Comment by Bowens20832** [comment]
@rabisnaqvi So Tina CMS to accomplish an admin portal?

Any other admin panels we could look at?

I was considering looking at Ghost's repo for some idea's as well: https://github.com/TryGhost/Ghost/tree/main/ghost/admin

**Comment by Bowens20832** [comment]
@rabisnaqvi I think I found a solution... How does this look?

![image](https://github.com/elegantframework/elegant-cli/assets/10189130/fbbbe6f8-4fd5-4bca-917e-bb9bc2a2d6f8)

**Reply by rabisnaqvi** [reply, marked as answer]
This is really good stuff, it looks amazing! I can see this can be really easy to use for people who already have experience with something like wordpress.

**Reply by Bowens20832** [reply]
@rabisnaqvi I finished putting together a rough functioning prototype this morning.

What do you think?

https://youtu.be/l6vIsmru34I

**Reply by rabisnaqvi** [reply]
@Bowens20832 this looks pretty cool! love it. also, I think we could probably handle the version controlling, like publishing changes, within the application's admin panel. what do you think.?

**Reply by Bowens20832** [reply]
@rabisnaqvi Yes I agree. The admin panel has a light connection to Git right now for publishing new content, but I believe we can continue to build on it and improve it more.

**Reply by Bowens20832** [reply]
@rabisnaqvi I also created a Discord chat for this project since we last talked. 

You can message me there or email me if you ever need feedback, but I'm not responding here fast enough.

https://discord.gg/PwY38x4uvV
